File: data-structures/segment-tree.cc
#inlcude "../header.h"


size_t ceilingTwo(size_t n) {
  size_t res = 1;
  while (res < n) {
    res <<= 1;
  }
  return res;
}

template<typename Segment>
class SegmentTree {
 public:
  // The root represent [0, n_).
  SegmentTree(size_t n) : n_(ceilingTwo(n)), segs_(n_ << 1) {
  }

  using Callback = std::function<bool(Segment &, size_t, size_t)>;
  using UpdateCallback =
      std::function<void(Segment &, const Segment &, const Segment &)>;

  // Traverse a disjoint set of segments exactly covering [l, r). The order is
  // from the leftmost to the rightmost.
  void traverseSegmentsFor(size_t l, size_t r, Callback& cb) {
    Callback overlapCheck = [&](Segment& seg, size_t sl, size_t sr) {
      if (sr <= l || r <= sl)  {
        return false;
      }
      if (l <= sl && sr <= r) {
        if (!cb(seg, sl, sr)) {
          return false;
        }
      }
      return true;
    };

    UpdateCallback null = [](Segment &, const Segment &, const Segment &) {

    };

    dfsIf(overlapCheck, null);
  }

  void dfsIf(Callback& cb, UpdateCallback& updateCb) {
    if (!cb(segs_[0], 0, n_)) {
      return;
    }
    dfsIf_(0, 0, n_, cb, updateCb);
  }

  void dfsIf_(size_t segIdx, size_t l, size_t r, Callback &cb,
              UpdateCallback &updateCb) {
    // std::cout << "seg @" << segIdx << ": [" << l << ", " << r << ")" << std::endl;
    if (l + 1 == r) {
      return;
    }

    size_t mid = (l + r) >> 1;

    const size_t lc = (segIdx << 1) + 1;
    const size_t rc = lc + 1;
    if (cb(segs_[lc], l, mid)) {
      dfsIf_(lc, l, mid, cb, updateCb);
    }

    if (cb(segs_[rc], mid, r)) {
      dfsIf_(rc, mid, r, cb, updateCb);
    }

    updateCb(segs_[segIdx], segs_[lc], segs_[rc]);
  }

  size_t n_;
  std::vector<Segment> segs_;
};

class SegmentTreeTest : public testing::Test {
 protected:
  void SetUp() override {}
  void TearDown() override {}
};


struct Seats {
  size_t minOccupied = 0;
  size_t sumOccupied = 0;
};

class BookMyShow {
public:
  BookMyShow(int n, int m) : m_(m), tree_(n) {}

  bool filterOutSeg(const Seats &seg, size_t l, size_t r, size_t maxR) {
    if (seg.minOccupied == m_) {
      return true;
    }
    if (maxR <= l) {
      // Not overlap
      return true;
    }
    return false;
  }

  std::vector<int> gather(int k, int maxRow) {
    int row = -1, col = -1;
    size_t maxR = maxRow + 1;
    SegmentTree<Seats>::Callback canGather = [&, this](Seats &seg, size_t l,
                                                       size_t r) -> bool {
      // std::cout << "[" << l << ", " << r << ") = " << seg.minOccupied << " @"
      // << (&seg - &tree_.segs_[0]) << std::endl;

      if (row >= 0) {
        // Already found a row
        return false;
      }
      if (filterOutSeg(seg, l, r, maxR)) {
        return false;
      }
      if (seg.minOccupied + k > m_) {
        // No enough capacity in the whole segment.
        return false;
      }
      if (l + 1 == r) {
        // std::cout << " Picked row " << l << std::endl;
        row = l;
        col = seg.minOccupied;
        seg.minOccupied += k;
        seg.sumOccupied += k;
      }
      return true;
    };

    // std::cout << " ====================== " << std::endl;
    tree_.dfsIf(canGather, updateCb_);
    if (row >= 0) {
      return {row, col};
    }
    return {};
  }

  bool scatter(int k, int maxRow) {
    bool updateNode = false;
    auto neededSeats = k;
    Tree::Callback scatterSeg = [&updateNode, &neededSeats, maxr = maxRow + 1,
                                 this](Seats &seg, size_t l, size_t r) -> bool {
      // std::cout << "[" << l << ", " << r << ") = " << seg.sumOccupied <<
      // std::endl;

      if (neededSeats == 0) {
        return false;
      }
      if (filterOutSeg(seg, l, r, maxr)) {
        return false;
      }

      if (maxr >= r) {
        // The rows in this segment are all usable.
        auto seats = (r - l) * m_ - seg.sumOccupied;
        if (seats <= neededSeats) {
          if (updateNode) {
            // The whole segment is occupied now.
            seg.minOccupied = m_;
            seg.sumOccupied = m_ * (r - l);
          }
          // std::cout << "    occupied " << seats << " seats." << std::endl;
          neededSeats -= seats;
          return false;
        } else if (l + 1 == r) {
          // Leaf node
          if (updateNode) {
            seg.minOccupied += neededSeats;
            seg.sumOccupied = seg.minOccupied;
          }
          // std::cout << "    occupied at a leaf " << neededSeats << " seats: "
          // << seg.minOccupied << " " << seg.sumOccupied << " @" << (&seg -
          // &tree_.segs_[0]) << std::endl;
          neededSeats = 0;
          return false;
        }
      }
      return true;
    };

    Tree::UpdateCallback nullCb = [](Seats &, const Seats &, const Seats &) {};
    // std::cout << " ====================== " << std::endl;
    tree_.dfsIf(scatterSeg, nullCb);
    if (neededSeats > 0) {
      return false;
    }

    neededSeats = k;
    updateNode = true;
    // std::cout << " ====================== " << std::endl;
    tree_.dfsIf(scatterSeg, updateCb_);
    return true;
  }

  void updateSegment(Seats &parent, const Seats &lc, const Seats &rc) {
    parent.minOccupied = std::min(lc.minOccupied, rc.minOccupied);
    parent.sumOccupied = lc.sumOccupied + rc.sumOccupied;
  }

  size_t m_;
  using Tree = SegmentTree<Seats>;
  Tree tree_;
  SegmentTree<Seats>::UpdateCallback updateCb_ =
      std::bind(&BookMyShow::updateSegment, this, std::placeholders::_1,
                std::placeholders::_2, std::placeholders::_3);
};

/**
 * Your BookMyShow object will be instantiated and called as such:
 * BookMyShow* obj = new BookMyShow(n, m);
 * vector<int> param_1 = obj->gather(k,maxRow);
 * bool param_2 = obj->scatter(k,maxRow);
 */

TEST_F(SegmentTreeTest, LeetCodeBookMyShow) {
  {
    BookMyShow bms(2, 5); // There are 2 rows with 5 seats each
    EXPECT_EQ(
        bms.gather(4, 0),
        std::vector<int>({0, 0})); // return [0, 0]
                                   // The group books seats [0, 3] of row 0.
    EXPECT_TRUE(
        bms.gather(2, 0)
            .empty()); // return []
                       // There is only 1 seat left in row 0,
                       // so it is not possible to book 2 consecutive seats.
    EXPECT_TRUE(bms.scatter(
        5, 1)); // return True
                // The group books seat 4 of row 0 and seats [0, 3] of row 1.

    EXPECT_FALSE(bms.scatter(5, 1)); // return False
  }
  {
    BookMyShow show(4, 5);
    EXPECT_TRUE(show.scatter(6, 2));
    EXPECT_TRUE(show.gather(6, 3).empty());
    EXPECT_FALSE(show.scatter(9, 1));
  }

  {

    BookMyShow show(5, 9);
    EXPECT_TRUE(show.gather(10, 1).empty());
    EXPECT_TRUE(show.scatter(3, 3));
    EXPECT_EQ(show.gather(9, 1), std::vector<int>({ 1, 0 }));
    EXPECT_EQ(show.gather(10, 2), std::vector<int>({ }));
    EXPECT_EQ(show.gather(2, 0), std::vector<int>({ 0, 3}));

  }
}

int main(int argc, char* argv[]) {
  testing::InitGoogleTest(&argc, argv);
  gflags::ParseCommandLineFlags(&argc, &argv, true);
  google::InitGoogleLogging(argv[0]);
  return RUN_ALL_TESTS();
}

File: python/segment-tree.py
from dataclasses import dataclass
from turtle import update
from xmlrpc.client import boolean


@dataclass
class Segment(object):
    l: int
    # Exclusive r > l
    r: int
    max_row_seats: int = 0
    sum_row_seats: int = 0
    left: object = None
    right: object = None


def printTree(seg, indents="  "):
    print(
        f"{indents}[{seg.l}, {seg.r}) = max({seg.max_row_seats}) sum({seg.sum_row_seats})"
    )
    if seg.left is not None:
        printTree(seg.left, indents + "  ")
    if seg.right is not None:
        printTree(seg.right, indents + "  ")


def constructSegmentTree(seg, m):
    mid = (seg.l + seg.r) // 2
    if seg.l == mid:
        return

    assert mid < seg.r and seg.l < mid
    l_seg = Segment(
        l=seg.l, r=mid, max_row_seats=m, sum_row_seats=m * (mid - seg.l)
    )
    r_seg = Segment(
        l=mid, r=seg.r, max_row_seats=m, sum_row_seats=m * (seg.r - mid)
    )

    constructSegmentTree(l_seg, m)
    constructSegmentTree(r_seg, m)
    seg.left = l_seg
    seg.right = r_seg


def updateSeg(seg):
    seg.max_row_seats = max(seg.left.max_row_seats, seg.right.max_row_seats)
    seg.sum_row_seats = seg.left.sum_row_seats + seg.right.sum_row_seats


def gather(seg, k, maxRow, m):
    assert maxRow >= seg.l
    if seg.max_row_seats < k:  # or seg.fully_occupied:
        return None
    if seg.left is None:
        seg.max_row_seats -= k
        seg.sum_row_seats -= k
        return [seg.l, m - seg.sum_row_seats - k]

    l_res = gather(seg.left, k, maxRow, m)
    if l_res is not None:
        updateSeg(seg)
        return l_res

    if maxRow >= seg.right.l:
        r_res = gather(seg.right, k, maxRow, m)
        if r_res is not None:
            updateSeg(seg)
            return r_res
    return None


def can_scatter(seg, k, maxRow):
    if seg.sum_row_seats < k:
        return False
    if seg.left is None:
        assert seg.sum_row_seats >= k
        return True

    if maxRow < seg.left.r:
        return can_scatter(seg.left, k, maxRow)
    elif k <= seg.left.sum_row_seats:
        return True
    else:
        return can_scatter(seg.right, k - seg.left.sum_row_seats, maxRow)


def scatter(seg, k, maxRow):
    if seg.left is None:
        assert k <= seg.sum_row_seats, (seg, k, maxRow)
        assert seg.l <= maxRow
        seg.sum_row_seats -= k
        seg.max_row_seats -= k
        return

    if k == seg.sum_row_seats:
        assert seg.r - 1 <= maxRow
        seg.max_row_seats = seg.sum_row_seats = 0
        return

    if maxRow < seg.left.r or k <= seg.left.sum_row_seats:
        scatter(seg.left, k, maxRow)
    else:
        l_seats = seg.left.sum_row_seats
        scatter(seg.left, l_seats, maxRow)
        scatter(seg.right, k - l_seats, maxRow)

    updateSeg(seg)


class BookMyShow(object):
    def __init__(self, n, m):
        """
        :type n: int
        :type m: int
        """
        self.root = Segment(
            l=0,
            r=n,
            max_row_seats=m,
            sum_row_seats=m * n,
        )
        self.m = m
        constructSegmentTree(self.root, m)

    def gather(self, k, maxRow):
        """
        :type k: int
        :type maxRow: int
        :rtype: List[int]
        """
        res = gather(self.root, k, maxRow, self.m)
        return res if res is not None else []

    def scatter(self, k, maxRow):
        """
        :type k: int
        :type maxRow: int
        :rtype: bool
        """
        if can_scatter(self.root, k, maxRow):
            scatter(self.root, k, maxRow)
            return True
        return False


bms = BookMyShow(2, 5)
print(bms.gather(4, 0))
print(bms.gather(2, 0))

print(bms.scatter(5, 1))
print(bms.scatter(5, 1))


bms = BookMyShow(4, 5)
print(bms.scatter(6, 2))

print(bms.gather(6, 3))

print(bms.scatter(9, 1))

print("----------")

bms = BookMyShow(5, 9)

print(bms.gather(10, 1))
printTree(bms.root)
print("=========")

print(bms.scatter(3, 3))
printTree(bms.root)
print("=========")

print(bms.gather(9, 1))
printTree(bms.root)
print("=========")

print(bms.gather(10, 2))
print(bms.gather(2, 0))


# Your BookMyShow object will be instantiated and called as such:
# obj = BookMyShow(n, m)
# param_1 = obj.gather(k,maxRow)
# param_2 = obj.scatter(k,maxRow)
